lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -205,116 +198,171 @@ def _on_list_backups_action(self, event: ActionEvent) -> None:
+            return False
+
+    def _on_restore_backup_action(self, event: ActionEvent) -> None:
         """Restores a backup to the current cluster."""
         if not self._can_unit_perform_backup(event):
             event.fail("Failed: backup service is not configured yet")
             return
-
+        if not self._is_restore_complete():
+            event.fail("Failed: previous restore is still in progress")
+            return
+        # Now, validate the backup is working
         backup_id = str(event.params.get("backup-id"))
+        if self._is_backup_available_for_restore(backup_id):
+            event.fail(f"Failed: no backup-id {backup_id}")
+            return
+
+        self.charm.status.set(MaintenanceStatus(RestoreStarting))
 
         # Restore will try to close indices if there is a matching name.
         # The goal is to leave the cluster in a running state, even if the restore fails.
         # In case of failure, then restore action must return a list of closed indices
         closed_idx = set()
         try:
-            backups = self._list_backups()
-            if backup_id not in backups.keys():
-                event.fail(f"Failed: no backup-id {backup_id}, options available: {backups}")
-                return
-            backup_id_state = self.get_snapshot_status(backups[backup_id]["state"])
-            if backup_id_state != BackupServiceState.SUCCESS:
-                event.fail(
-                    f"Failed: no backup-id {backup_id} not successful, state is: {backup_id_state}"
-                )
-                return
-            output, closed_idx = self._restore_and_try_close_indices_if_needed(backup_id)
+            closed_idx = self._close_indices_if_needed(backup_id)
+            output = self._restore(backup_id)
             logger.debug(f"Restore action: received response: {output}")
             logger.info(f"Restore action succeeded for backup_id {backup_id}")
-        except OpenSearchListBackupError as e:
-            event.fail(f"Failed: {e}, closed indices: {closed_idx}")
+        except (
+            OpenSearchHttpError,
+            OpenSearchRestoreIndexClosingError,
+            OpenSearchRestoreCheckError,
+        ) as e:
+            event.fail(f"Failed: {e}")
+            self.charm.status.set(BlockedStatus(RestoreActionFailed))
             return
-        if self.get_service_status(output) != BackupServiceState.SUCCESS:
-            status = str(self.get_service_status(output))
-            event.fail(f"Failed with: {status}, closed indices: {closed_idx}")
-        event.set_results({"state": "successful restore!", "closed-indices": str(closed_idx)})
 
-    def _on_create_backup_action(self, event: ActionEvent) -> None:
+        # Post execution checks
+        # Was the call successful?
+        state = self.get_service_status(output)
+        if state != BackupServiceState.SUCCESS:
+            self.charm.status.set(BlockedStatus(RestoreActionFailed))
```

Comment:
the status should not be altered on a failure, `event.fail` should be used here.